storage/flush.go

```diff
@@ -35,63 +34,94 @@ import (
 )
 
 var (
-	errFlushAlreadyInProgress = errors.New("flush already in progress")
+	errFlushOrSnapshotAlreadyInProgress = errors.New("flush or snapshot already in progress")
 )
 
 type flushManager struct {
 	sync.RWMutex
 
-	database        database
-	opts            Options
-	nowFn           clock.NowFn
-	pm              persist.Manager
-	flushInProgress bool
-	status          tally.Gauge
+	database database
+	opts     Options
+	pm       persist.Manager
+	// isFlushingOrSnapshotting is used to protect the flush manager against
+	// concurrent use, while flushInProgress and snapshotInProgress are more
+	// granular and are used for emitting granular gauges.
+	isFlushingOrSnapshotting bool
+	flushInProgress          bool
+	snapshotInProgress       bool
```

Comment:
Hm can we combine these three bools into a single enum?  Otherwise there's states that are invalid and two variables controlling the same overall derived state which is never good (always avoid multiple sources of truth), etc.

i.e.
```go
type flushManagerState int

const (
  flushManagerIdle flushManagerState = iota
  flushManagerFlushInProgress
  flushManagerSnapshotInProgress
)
```

Reply:
good call, done. Although I had to introduce one additional possible state:

```golang
const (
	flushManagerIdle flushManagerState = iota
	// flushManagerNotIdle is used to protect the flush manager from concurrent use
	// when we haven't begun either a flush or snapshot.
	flushManagerNotIdle
	flushManagerFlushInProgress
	flushManagerSnapshotInProgress
)
```